### PR TITLE
SWARM-1327: Approach 3. Explode jar files belonging to boot loader resource root into directories.

### PR DESCRIPTION
**Pull request by juangon**
Let's see the difference in tests executing between this approach and previous one due to the exploding.

Main differences are in changes made to core/boostrap...ApplicationModuleFinder and NestedJarResourceLoader.

### CODE REVIEW

File: core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/util/BootstrapUtil.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            if (je.isDirectory()) {
+                continue;
+            }
+            InputStream is = null;
+            FileOutputStream fo = null;
+            try {
+                is = jarFile.getInputStream(je);
+                fo = new FileOutputStream(fl);
+                while (is.available() > 0) {
+                    fo.write(is.read());
```
Comment (Ladicek) [inline core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/util/BootstrapUtil.java]: Good for testing, but one byte at a time will be dog slow :-)

File: core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/modules/NestedJarResourceLoader.java [hunk 2]
```diff
@@ -131,23 +132,42 @@ public static ResourceLoader loaderFor(URL base, String rootPath, String loaderP
...
             }
         } else if (urlString.startsWith("file:")) {
             if (loaderName.endsWith(".jar") || loaderName.endsWith(".war")) {
-                return ResourceLoaders.createJarResourceLoader(
-                        loaderName,
-                        new JarFile(new File(urlString.substring(5), loaderPath)));
+                final File file = new File(urlString.substring(5), loaderPath);
+                final JarFile jarFile = new JarFile(file);
+
+                File tmpDir = TempFileManager.INSTANCE.newTempDirectory("nextedjarloader", null);
```
Comment (Ladicek) [inline core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/modules/NestedJarResourceLoader.java]: `nexted` :-)

**Comment by Ladicek** [comment]
Wow, it passed even on Windows? That's cool. Are we sure that the temporary exploded directories would be caught by the enforcer as well? The result is very nice!

I'm still not sure about the `stopMain` thing, it's too intrusive to me. But maybe there's no better way, I don't know.

File: core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/util/BootstrapUtil.java (hunk 1 above)
Comment (juangon) [reply core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/util/BootstrapUtil.java]: Sorry, too much copy/paste here :-)

**Comment by juangon** [comment]
Yes @Ladicek it is also caught by enforcer plugin as it accomplish the rule of 5 numbers or more in the name (regardless if it's directory or not).

**Comment by juangon** [comment]
FYI the only file leak check I had to disable was testsuite-webservices. It seems caused by server not releasing the provided HandledChain xml file on shutdown in Windows, which prevent files to be deleted.

**Review by bobmcwhirter — APPROVED** [review]

File: core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/Main.java [hunk 3]
```diff
@@ -34,16 +55,77 @@ public static void main(String... args) throws Throwable {
...
+            String processFile = System.getProperty(MAIN_PROCESS_FILE);
+
+            if (processFile != null) {
+                shutdownService = Executors.newSingleThreadExecutor();
+                shutdownService.submit(() -> {
+                        File uuidFile = new File(processFile);
+                        try {
+                            File watchedDir = uuidFile.getParentFile();
+                            register(watchedDir);
+                            processEvents(watchedDir, uuidFile.toPath());
```
Comment (Ladicek) [inline core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/Main.java:68]: I'd inline the `register` method, it's just 2 lines, and mainly, the `processEvents` should be called something like `waitForFileDeleted` or something like that.

**Comment by Ladicek** [comment]
OK, just one minor comment, otherwise LGTM.

**Comment by juangon** [comment]
Oh I was going to say tests are almost spending same time, so good to go :-)

**Pull request merged**